src/protocol/boolean/bit_decomposition.rs

```diff
@@ -38,49 +40,170 @@ impl BitDecomposition {
+            p_minus_c.into(),
         )
         .await?;
 
-        // Step 7. a bitwise scalar value `f_B = bits(2^el - p)`
+        // Step 4 has a lot going on. We'd going to break it down into substeps.
+        // Step 4.1. Make a bitwise scalar value of f = 2^el + c - p.
+        // let el = usize::try_from(u128::BITS - F::PRIME.into().leading_zeros()).unwrap();
         let el = u128::BITS - F::PRIME.into().leading_zeros();
-        let x = (1 << el) - F::PRIME.into();
+        let two_exp_el = u128::pow(2, el);
+        let _f_int: u128 = two_exp_el + c.as_u128() - F::PRIME.into();
+        debug_assert!(_f_int < u64::max_value().into());
+        let f_int: u64 = _f_int as u64;
+
+        // Step 4.2. Compute [g]_B = (f_i - c_i) [q]_p + c_i
+        let mut g_bin: Vec<G> = Vec::with_capacity(usize::try_from(el).unwrap());
+        for bit_index in 0..el {
+            // these are single bits, so let's make them bools
+            let f_i: bool = (f_int >> bit_index) & 1 == 1;
+            let c_i: bool = (c.as_u128() >> bit_index) & 1 == 1;
+            // g_i can either be c_i (known to be either 0, 1), or either [q]_p, ¬[q]_p
+            // using an enum here because we're going to do something cleaver in bitwise add
+            let g_i: G = match (f_i, c_i) {
+                // Case where f_i - c_i == 0
+                (false, false) | (true, true) => match c_i {
+                    true => G::One,
+                    false => G::Zero,
+                },
+                // Case where f_i - c_i = 1
+                (true, false) => G::Q,
+                // Case where f_i - c_i = -1
+                (false, true) => G::NotQ,
+                // this is what g_i should actually be here, so we can use this down the road
+                // let g_i_foo = S::share_known_value(&ctx, F::ONE) - &q_p.clone();
+            };
+            g_bin.push(g_i);
+            // let f_i_minus_c_i: u64 = f_i - c_i;
+            // let g_prime: S = q_p.clone().into() * f_i_minus_c_i;
+            // // c_i ∈ {0, 1}, so F::truncate_from will work
+            // debug_assert!(c_i <= 1_u128);
+            // g_bin.push(S::share_known_value(&ctx, F::truncate_from(c_i)) + &g_prime);
+        }
+
+        // Step 5. Compute BitwiseSum([r]_B, [g]_B)
+        let mut h: Vec<S> = Vec::with_capacity(usize::try_from(el).unwrap());
+        let one_minus_q_p = S::share_known_value(&ctx, F::ONE) - &q_p.clone();
+        let (mut last_carry, result_bit, mut last_carry_known_to_be_zero) = match g_bin[0] {
+            G::Zero => (S::ZERO, r.b_b[0].clone(), true),
+            G::One => (r.b_b[0].clone(), S::share_known_value(&ctx, F::ONE), false),
+            G::Q => (
+                q_p.multiply(&r.b_b[0], ctx.narrow(&Step::AddGtoR), record_id)
+                    .await?,
+                q_p.clone() + &r.b_b[0],
+                false,
+            ),
+            G::NotQ => (
+                one_minus_q_p
+                    .clone()
```

Comment:
```suggestion
```

Reply:
what should I read to know when to clone and when not to?

Reply:
also, I'm getting an error that this value has been moved, and it was moved here.

Reply:
> what should I read to know when to clone and when not to?

Rust standard library docs are pretty good:
https://doc.rust-lang.org/std/clone/trait.Clone.html
https://doc.rust-lang.org/std/marker/trait.Copy.html

Reply:
>also, I'm getting an error that this value has been moved, and it was moved here.

it is weird because `multiply` should take it by reference

Reply:
our code is a little weird in a few ways.

We don't implement `Copy` for shares when we probably could; that is, we require that you explicitly call `clone()` so that the cost of creating a new instance is explicit.

We often do require a new instance: most operations return a new value, so you need the output to be a new instance; but we often only have references (because we want to avoid making copies).  Also, our operations generally have a full instance on the left and a reference on the right.  That seems a bit weird until you realize that what that allows us to do is have the left operand be modified by the various operations and returned as the output.